pkg/inject/inject.go

```diff
@@ -0,0 +1,594 @@
+}
+
+// ResourceConfig contains the parsed information for a given workload
+type ResourceConfig struct {
+	globalConfig          *config.Global
+	proxyConfig           *config.Proxy
+	nsAnnotations         map[string]string
+	meta                  metav1.TypeMeta
+	obj                   runtime.Object
+	objMeta               objMeta
+	podLabels             map[string]string
+	podSpec               *v1.PodSpec
+	dnsNameOverride       string
+	proxyOutboundCapacity map[string]uint
+}
+
+// NewResourceConfig creates and initializes a ResourceConfig
+func NewResourceConfig(globalConfig *config.Global, proxyConfig *config.Proxy) *ResourceConfig {
+	return &ResourceConfig{
+		globalConfig:          globalConfig,
+		proxyConfig:           proxyConfig,
+		podLabels:             map[string]string{k8s.ControllerNSLabel: globalConfig.GetLinkerdNamespace()},
+		proxyOutboundCapacity: map[string]uint{},
+	}
+}
+
+// WithMeta enriches ResourceConfig with extra metadata
+func (conf *ResourceConfig) WithMeta(kind, namespace, name string) *ResourceConfig {
+	conf.meta = metav1.TypeMeta{Kind: kind}
+	conf.objMeta = objMeta{&metav1.ObjectMeta{Name: name, Namespace: namespace}}
+	return conf
+}
+
+// WithNsAnnotations enriches ResourceConfig with the namespace annotations, that can
+// be used in shouldInject()
+func (conf *ResourceConfig) WithNsAnnotations(m map[string]string) *ResourceConfig {
+	conf.nsAnnotations = m
+	return conf
+}
+
+// WithProxyOutboundCapacity enriches ResourceConfig with a map of image names
+// to capacities, which can be used by the install code to modify the outbound
+// capacity for the prometheus container in the control plane install
+func (conf *ResourceConfig) WithProxyOutboundCapacity(m map[string]uint) *ResourceConfig {
+	conf.proxyOutboundCapacity = m
+	return conf
+}
+
+// YamlMarshalObj returns the yaml for the workload in conf
+func (conf *ResourceConfig) YamlMarshalObj() ([]byte, error) {
+	return yaml.Marshal(conf.obj)
+}
+
+// ParseMetaAndYaml fills conf fields with both the metatada and the workload contents
+func (conf *ResourceConfig) ParseMetaAndYaml(bytes []byte) (*Report, error) {
+	if _, err := conf.ParseMeta(bytes); err != nil {
+		return nil, err
+	}
+	r := newReport(conf)
+	return &r, conf.parse(bytes)
+}
+
+// ParseMeta extracts metadata from bytes.
+// It returns false if the workload's payload is empty
+func (conf *ResourceConfig) ParseMeta(bytes []byte) (bool, error) {
+	if err := yaml.Unmarshal(bytes, &conf.meta); err != nil {
+		return false, err
+	}
+	if err := yaml.Unmarshal(bytes, &conf.objMeta); err != nil {
+		return false, err
+	}
+	return conf.objMeta.ObjectMeta != nil, nil
+}
+
+// GetPatch returns the JSON patch containing the proxy and init containers specs, if any
+func (conf *ResourceConfig) GetPatch(bytes []byte) ([]byte, []Report, error) {
+	report := newReport(conf)
+	log.Infof("working on %s %s..", strings.ToLower(conf.meta.Kind), conf.objMeta.Name)
```

Comment:
I noticed that this log line is printed as:

```
time="2019-03-01T21:39:06Z" level=info msg="working on deployment .."
```

Looks like the proxy injector is calling `WithMeta` with these fields:

```
WithMeta(request.Kind.Kind, request.Namespace, request.Name)
```

But I think the issue is that `request.Namespace` and `request.Name` could be blank, based on this documentation:

```
	// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
	// rely on the server to generate the name.  If that is the case, this method will return the empty string.
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,5,opt,name=name"`
	// Namespace is the namespace associated with the request (if any).
	// +optional
	Namespace string `json:"namespace,omitempty" protobuf:"bytes,6,opt,name=namespace"`
```

The previous implementation parsed `request.Object.Raw` as a Deployment object and used the deployment's name instead of `request.Name`. But if you took that approach, then you'd end up parsing the object meta twice -- once in webhook.go, and once in the `parse` func in this file.

Off the top of my head I can't think of a great way to fix this. If you moved the log line after the `parse` call, that would work, but it also suggests that we're not attaching the right metadata when calling `WithMeta`.

Reply:
@klingerf We also parse the meta twice in CLI inject (even before this PR) for the very purpose of extracting the Kind and Name. Given the `AdmissionRequest`'s Name is useless, I'll pre-parse `request.Object.Raw` as well to extract it.
I was gonna make a PR for this but I pushed it by mistake to this branch :man_shrugging: